torch/csrc/jit/codegen/cuda/executor.cpp

```diff
@@ -529,15 +541,26 @@ LaunchParams FusionExecutor::computeLaunchParams(
   const uint64_t static_smem_size =
       computeSharedMemory(expr_eval, kernel_summary.static_smem_allocations);
 
-  TORCH_INTERNAL_ASSERT(
-      (dynamic_smem_size + static_smem_size) < max_device_smem,
-      "The total shared memory allocation is larger than available memory.",
-      " Dynamic size: ",
-      dynamic_smem_size,
-      ". Static size: ",
-      static_smem_size,
-      ". Available size: ",
-      max_device_smem);
+  // Check that requested smem size can be dynamically allocated.
+  //  This check is only done once a kernel has been compiled, since
+  //  maybe_available_dynamic_smem_ needs to be evaluated on
+  //  a compiled kernel.
+  if (maybe_available_dynamic_smem_.has_value()) {
+    maybe_available_dynamic_smem_ = maybe_available_dynamic_smem_.value();
+    TORCH_INTERNAL_ASSERT(
+        (dynamic_smem_size) < (maybe_available_dynamic_smem_.value() +
+                               device_smem_limit_ - configured_device_smem_),
```

Comment:
This seems a little convoluted to me, why isn't this not just `dynamic_smem_size < device_smem_limit_` ?

Reply:
The extra complication comes from `maybe_available_dynamic_smem_`, and `configured_device_smem_` from driver api. 

basically we want to supply `dynamic_smem_size` with two possible components of memory space:

`maybe_available_dynamic_smem_` = amount of memory we still have in `configured_device_smem_` (static sharedmem, aka. magic zero only at this point, already subtracted in this value)

and

`device_smem_limit_ - configured_device_smem_` = max additional shared memory we can request from L1.

Reply:
Maybe it would be more clear if you describe all variables in a comment, maybe just `maybe_available_dynamic_smem`.

Reply:
renamed and added comments. Thanks.